lang/java/avro/src/main/java/org/apache/avro/SchemaCompatibility.java

```diff
@@ -408,11 +390,30 @@ private SchemaCompatibilityType calculateCompatibility(
    * Identifies the type of a schema compatibility result.
    */
   public static enum SchemaCompatibilityType {
-    COMPATIBLE,
-    INCOMPATIBLE,
```

Comment:
This is a breaking change and would require holding this back until at least 1.8. I don't see a need to make a breaking change, though. What about adding a version of `getCompatibility` that returns a `SchemaPairCompatibility` instead of the `SchemaCompatibilityType`? Then you could fill in the specific message (with even more detail, like record field path) and return all the way up for the same result.

You'd probably want an internal method to do the check and memoize results that `getCompatibility` calls and retrieves the compatibility type result from. `calculateCompatibility` could be used for that purpose, but I think it would be cleaner to memoize separately in a `memoizeCalculateCompatibility`. You could return the correct `SchemaPairCompatibility` from `checkReaderWriterCompatibility`. Does that sound reasonable? Then this could easily be a compatible change.